File: solutions/cpp/Compare_k-th_largest_in_heap.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <cassert>
#include <iostream>
#include <vector>

using std::cout;
using std::endl;
using std::vector;

void CompareKthLargestHeapHelper(const vector<int>& max_heap, int k, int x,
                                 int idx, int* larger_x, int* equal_x);

// @include
typedef enum { SMALLER, EQUAL, LARGER } Ordering;

Ordering CompareKthLargestHeap(const vector<int>& max_heap, int k, int x) {
  int larger_x = 0, equal_x = 0;
  CompareKthLargestHeapHelper(max_heap, k, x, 0, &larger_x, &equal_x);
  return larger_x >= k ? LARGER : (larger_x + equal_x >= k ? EQUAL : SMALLER);
}

void CompareKthLargestHeapHelper(const vector<int>& max_heap, int k, int x,
                                 int idx, int* larger_x, int* equal_x) {
  // Check early termination.  Note that we cannot early terminate for
  // equal_x >= k because larger_x (which is currently smaller than k)
  // may still become >= k.
  if (*larger_x >= k || idx >= max_heap.size() || max_heap[idx] < x) {
    return;
  } else if (max_heap[idx] == x) {
    if (++*equal_x >= k) {
      return;
    }
  } else {  // max_heap[idx] > x.
    ++*larger_x;
  }
  CompareKthLargestHeapHelper(max_heap, k, x, 2 * idx + 1, larger_x, equal_x);
  CompareKthLargestHeapHelper(max_heap, k, x, 2 * idx + 2, larger_x, equal_x);
}
// @exclude

void SmallTest() {
  vector<int> max_heap = {10, 2, 9, 2, 2, 8, 8, 2, 2, 2, 2, 7, 7, 7, 7};
  assert(LARGER == CompareKthLargestHeap(max_heap, 3, 2));  // kth is larger.
  assert(LARGER == CompareKthLargestHeap(max_heap, 6, 2));  // kth is larger.
}

int main(int argc, char* argv[]) {
  SmallTest();
  //      5
  //    4  5
  //  4  4 4 3
  // 4
  vector<int> max_heap = {5, 4, 5, 4, 4, 4, 3, 4};
  int k, x;
  if (argc == 3) {
    k = atoi(argv[1]), x = atoi(argv[2]);
    Ordering res = CompareKthLargestHeap(max_heap, k, x);
    cout << (res == SMALLER ? "smaller" : (res == EQUAL ? "equal" : "larger"))
         << endl;
  } else {
    assert(SMALLER ==
           CompareKthLargestHeap(max_heap, 1, 6));  // expect smaller
    assert(EQUAL == CompareKthLargestHeap(max_heap, 1, 5));  // expect equal
    assert(EQUAL == CompareKthLargestHeap(max_heap, 6, 4));  // expect equal
    assert(EQUAL == CompareKthLargestHeap(max_heap, 3, 4));  // expect equal
    assert(SMALLER ==
           CompareKthLargestHeap(max_heap, 8, 4));  // expect smaller
    assert(LARGER == CompareKthLargestHeap(max_heap, 2, 4));  // expect larger
    assert(LARGER == CompareKthLargestHeap(max_heap, 2, 3));  // expect larger
  }
  return 0;
}

File: solutions/python/compare_k_th_largest_in_heap.py
import sys

# @include
SMALLER, EQUAL, LARGER = range(3)


def compare_kth_largest_heap(max_heap, k, x):
    larger_x = equal_x = 0
    larger_x, equal_x = compare_kth_largest_heap_helper(max_heap, k, x, 0,
                                                        larger_x, equal_x)
    return LARGER if larger_x >= k else (EQUAL if larger_x + equal_x >= k else
                                         SMALLER)


def compare_kth_largest_heap_helper(max_heap, k, x, idx, larger_x, equal_x):
    # Check early termination. Note that we cannot early terminate for
    # equal_x >= k because larger_x (which is currently smaller than k)
    # may still become >= k.
    if larger_x >= k or idx >= len(max_heap) or max_heap[idx] < x:
        return larger_x, equal_x
    elif max_heap[idx] == x:
        equal_x += 1
        if equal_x >= k:
            return larger_x, equal_x
    else:  # max_heap[idx] > x.
        larger_x += 1
    larger_x, equal_x = compare_kth_largest_heap_helper(
        max_heap, k, x, 2 * idx + 1, larger_x, equal_x)
    larger_x, equal_x = compare_kth_largest_heap_helper(
        max_heap, k, x, 2 * idx + 2, larger_x, equal_x)
    return larger_x, equal_x
# @exclude


def small_test():
    max_heap = [10, 2, 9, 2, 2, 8, 8, 2, 2, 2, 2, 7, 7, 7, 7]
    assert LARGER == compare_kth_largest_heap(max_heap, 3, 2)
    assert LARGER == compare_kth_largest_heap(max_heap, 6, 2)


def main():
    small_test()
    #     5
    #   4   5
    #  4 4 4 3
    # 4
    max_heap = [5, 4, 5, 4, 4, 4, 3, 4]
    if len(sys.argv) == 3:
        k = int(sys.argv[1])
        x = int(sys.argv[2])
        res = compare_kth_largest_heap(max_heap, k, x)
        print('smaller' if res == SMALLER else ('equal'
                                                if res == EQUAL else 'larger'))
    else:
        assert SMALLER == compare_kth_largest_heap(max_heap, 1, 6)
        assert EQUAL == compare_kth_largest_heap(max_heap, 1, 5)
        assert EQUAL == compare_kth_largest_heap(max_heap, 6, 4)
        assert EQUAL == compare_kth_largest_heap(max_heap, 3, 4)
        assert SMALLER == compare_kth_largest_heap(max_heap, 8, 4)
        assert LARGER == compare_kth_largest_heap(max_heap, 2, 4)
        assert LARGER == compare_kth_largest_heap(max_heap, 2, 3)


if __name__ == '__main__':
    main()
